### PR TITLE
Move rate-limiting from per-client to per-client-per-domain

### PR DESCRIPTION
**Pull request by DL6ER**
# What does this implement/fix?

Let me be clear here: Rate-limit is there it *protect* you. It is not meant as a method to cause pain. What are we protecting against? Basically against [Denial of Service (DoS)](https://de.wikipedia.org/wiki/Denial_of_Service) "attacks" either by a malicious attacker or by some client that has a bug or crashed or ... (you never know, we have seen so may things). If one attacker or or misbehaving client can take down your entire network because your DNS resolver runs out of memory - that doesn't sound like something you'd want, right? - and this is where FTL's rate-limiting kicks in, simply discarding queries when there are too many. The default settings are suitable both for home and enterprise, small and huge networks, however, they can always be adjusted (or even disabled) to your liking if you feel you need no DoS-protection.

This PR proposes to change FTL's rate-limiting mechanism from a pure "per-client" to a "per-client *and* per-domain" scheme. If a client queries `domain.com` like hell, it will get rate-limited as before. However, any *other* domain (say `example.com`) can still resolved (except, obviously, it it gets queried too often, too).

This reduces the "there is not Internet connection" effect some users reported at the costs of a much less effective rate-limiting as clients and/or malicious attackers can simply evade it by querying varying domains. We have seen this by buggy clients querying seemingly random subdomain - each of them would be counted on their own.

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** ***to be done***

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

### CODE REVIEW

**Comment by DL6ER** [comment]
Rebased on latest `development`

File: src/database/message-table.c [hunk 1]
```diff
@@ -26,8 +26,38 @@
...
+	{
+	case REGEX_MESSAGE:
+		return "REGEX";
+	case SUBNET_MESSAGE:
+		return "SUBNET";
+	case HOSTNAME_MESSAGE:
+		return "HOSTNAME";
+	case DNSMASQ_CONFIG_MESSAGE:
+		return "DNSMASQ_CONFIG";
+	case RATE_LIMIT_MESSAGE:
```
Comment (yubiuser) [inline src/database/message-table.c:41]: Why do we keep this here? `RATE_LIMIT_MESSAGE` is not used anymore.

File: src/dnsmasq_interface.c [hunk 2]
```diff
@@ -1550,17 +1553,51 @@ bool _FTL_CNAME(const char *dst, const char *src, const int id, const char* file
...
+	clientsData *client = getClient(clientID, true);
+	if(client == NULL)
+	{
+		logg("WARN: No memory available, skipping query analysis (%s:%d)",
+		     short_path(__FILE__), __LINE__);
+		free(child_domain_str);
+		return false;
+	}
+
+	// Get child's DNS cache entry (we eed to create a new cache entry here
```
Comment (yubiuser) [inline src/dnsmasq_interface.c:1587]:
```suggestion
	// Get child's DNS cache entry (we need to create a new cache entry here
```

File: src/enums.h [hunk 3]
```diff
@@ -211,6 +211,7 @@ enum message_type {
 	SHMEM_MESSAGE,
```
Comment (yubiuser) [inline src/enums.h:211]: Still need L208 "RATE_LIMIT_MESSAGE"?

**Comment by DL6ER** [comment]
The `unique` change we added recently won't work here as uniqueness is now composed out of three fields for this new message type (type + client + domain). Needs more work.

**Pull request closed without merging**